chain/events/filter/index_test.go

```diff
@@ -957,3 +960,87 @@ func TestEventIndexPrefillFilterExcludeReverted(t *testing.T) {
 		})
 	}
 }
+
+// TestQueryPlan is to ensure that future modifications to the db schema, or future upgrades to
+// sqlite, do not change the query plan of the prepared statements used by the event index such that
+// queries hit undesirable indexes which are likely to slow down the query.
+// Changes that break this test need to be sure that the query plan is still efficient for the
+// expected query patterns.
+func TestQueryPlan(t *testing.T) {
+	ei, err := NewEventIndex(context.Background(), filepath.Join(t.TempDir(), "actorevents.db"), nil)
+	require.NoError(t, err, "create event index")
+
+	verifyQueryPlan := func(stmt string) {
+		rows, err := ei.db.Query("EXPLAIN QUERY PLAN " + strings.Replace(stmt, "?", "1", -1))
+		require.NoError(t, err, "explain query plan for query: "+stmt)
+		defer func() {
+			require.NoError(t, rows.Close())
+		}()
+		// First response to EXPLAIN QUERY PLAN should show us the use of an index that we want to
+		// encounter first to narrow down the search space - either a height or tipset_key_cid index
+		// - sqlite_autoindex_events_seen_1 is for the UNIQUE constraint on events_seen
+		// - events_seen_height and events_seen_tipset_key_cid are explicit indexes on events_seen
+		// - event_height and event_tipset_key_cid are explicit indexes on event
+		rows.Next()
+		var id, parent, notused, detail string
+		require.NoError(t, rows.Scan(&id, &parent, &notused, &detail), "scan explain query plan for query: "+stmt)
+		detail = strings.TrimSpace(detail)
+		var expectedIndexes = []string{
+			"sqlite_autoindex_events_seen_1",
+			"events_seen_height",
+			"events_seen_tipset_key_cid",
+			"event_height",
+			"event_tipset_key_cid",
+		}
+		indexUsed := false
+		for _, index := range expectedIndexes {
+			if strings.Contains(detail, " INDEX "+index) {
+				indexUsed = true
```

Comment:
@rvagg But what if the detail uses some other index before using an expected index ? We wont fail the test here, right ?

Reply:
no, because we only pull the first row of the results - `row.Next()` above, subsequent rows contain the additional indexes